Enhance API-Testing-Guidelines.md

jlurien opened the pull request:
#### What type of PR is this?

* documentation

#### What this PR does / why we need it:

Enhances API testing guidelines with much more details and explanations. Initial version was too basic to assure consistent and coherent testing plans.

#### Which issue(s) this PR fixes:

Fixes #158 

#### Special notes for reviewers:

This is a first draft following proposal and initial feedback in  #158. 

#### Changelog input

```
* Enhancement of API-Testing-Guidelines

```

PedroDiez commented on documentation/API-Testing-Guidelines.md:
```diff
@@ -1,40 +1,284 @@
+
+### Request sending <a name="request-sending"></a>
+
+For scenarios testing individual operations, the request will be send by means of a single `When` step, indicating the `operationId` specified in the spec, as the `operationId` uniquely identifies the HTTP method and resource path. E.g.:
```
```suggestion
For scenarios testing individual operations, the request will be sent by means of a single `When` step, indicating the `operationId` specified in the spec, as the `operationId` uniquely identifies the HTTP method and resource path. E.g.:
```

rartych commented on documentation/API-Testing-Guidelines.md:
```diff
@@ -1,40 +1,284 @@
 This document captures guidelines for the API testing in CAMARA project. These guidelines are applicable to every API to be worked out under the CAMARA initiative.
 
 ## Test Cases contribution for APIs <a name="contribution"></a>
 Based on the decision taken in the Commonalities working group and as documented in [API-Readiness-Checklist](https://github.com/camaraproject/Commonalities/blob/main/documentation/API-Readiness-Checklist.md), at least one Gherkin feature file will be added to the main subproject repo and this will fulfill the minimum criteria of readiness with respect to API test cases and documentation. If subp ...
```
The updated API Readiness Checklist is proposed in Release Msanagement: https://github.com/camaraproject/ReleaseManagement/pull/29

It impacts this section and possibly "Basic API test cases & documentation" and "Enhanced API test cases & documentation"  should be defined here.

jlurien replied:
It's difficult to set the difference between Basic and Enhanced. To me basic tests cases should cover at least that the implementation interface is compliant with the API spec, in terms of syntax and support for all mandatory aspects. Enhanced test plan may validate unusual error scenarios and corner cases. But this is very subtle.

jlurien commented on documentation/API-Testing-Guidelines.md:
```diff
@@ -1,40 +1,284 @@
+
+- APIs with a single feature file will use the api-name as file name: E.g. `location-verification.feature`
+- For APIs with several operations that split the test plan in one feature file per operation, recommendation is to use the `operationId` as file name: `createSession.feature`.
+- For APIs with several feature files, not split by operation, may use the api-name along some other description, e.g.: `qod-notifications.feature`
```
Split is right I think :)

jlurien commented:
In the latter commit:

Proposal updated with comments and feedback after dedicated session on May 16th.

- Tests coverage section moved to to the top, merging legacy section "Test Cases contribution for APIs".
- Added distinction between basic and enhanced test plan, to be aligned with Release Management proposal. There is no proper definition of what is basic or enhanced, so I suggested some initial test.
- Added explicitly that environment variables may be kept in a separated file. We should enhance this part with more details and a suggested template.

Please review these changes.

rartych commented on documentation/API-Testing-Guidelines.md:
````diff
@@ -1,41 +1,288 @@
-Subproject_Repository/code/Test_definitions/file_name.feature
-for e.g. https://github.com/camaraproject/QualityOnDemand/blob/main/code/Test_definitions/QoD_API_Test.feature
-```
+Based on the decision taken in the Release Management and Commonalities working groups, and as documented in [API-Readiness-Checklist](https://github.com/camaraproject/Commonalities/blob/main/documentation/API-Readiness-Checklist.md), at least one Gherkin feature file will be added to the main subproject repo and this will fulfill the minimum criteria of readiness with respect to API test cases a ...
````
I propose to change the link to Release Management wiki.
Current API Readiness Checklist will be deprecated: https://github.com/camaraproject/Commonalities/issues/202
```suggestion
Based on the decision taken in the Release Management and Commonalities working groups, and as documented in [API Release Process](https://wiki.camaraproject.org/display/CAM/API+Release+Process)](https://wiki.camaraproject.org/display/CAM/API+Release+Process#APIReleaseProcess-APIreadinesschecklist), at least one Gherkin feature file will be added to the main subproject repo and this will fulfill the minimum criteria of readiness with respect to API test cases and documentation:
```

mdomale commented on documentation/API-Testing-Guidelines.md:
```diff
@@ -1,41 +1,288 @@
+
+### Independent operations <a name="independent-operations"></a>
+
+Operations that execute an action or retrieve information from the operator, without modifying a resource, will be tested in its own dedicated feature file. 
```
Does this mean we will need two different feature files for QoD Operations ? If yes then what would be significance of it ?

jlurien replied:
QoD is mostly a CRUD API, encapsulating sessions resources, so it lies on the other category. Anyway, it is up to QoD subproject, and the contributors of its test plans, to decide on how many feature files are convenient. Splitting a test plan in several features is a test plan design decision, taking into account the complexity of the test plan. The more complex a test plan is, the most likely is that it will be split into several feature files. As in any software project.

These guidelines are recommendations for the design, so we may make the text less normative: "will be tested in its own dedicated feature file" -> "will be **usually** tested in its own dedicated feature file"

mdomale replied:
@jlurien  OK with us kindly accommodate this in PR file as well

jlurien replied:
Added

mdomale commented on documentation/API-Testing-Guidelines.md:
```diff
@@ -1,41 +1,288 @@
+For operations receiving an input request body with optional properties, several scenarios should be included, testing different sets of valid inputs.
+
+
+### CRUD operations <a name="crud-operations"></a>
```
Can we rename this to Resource Modifying Operations to make the differentiation clearer

jlurien replied:
Fine to me

mdomale replied:
@jlurien  Kindly accommodate in PR file as well

mdomale commented on documentation/API-Testing-Guidelines.md:
```diff
@@ -1,41 +1,288 @@
+
+### CRUD operations <a name="crud-operations"></a>
+
+For path resources with CRUD (Create/Read/Update/Delete) operations, some scenario should validate that the new state of the resource is coherent with the response to the operation.
```
* response -> request correction

jlurien replied:
I can clarify this further, but it really meant the **response**. The request may ask the server for something, but it is the response which tells the client whether the server has granted the request and therefore modify the state of the resource accordingly.

mdomale replied:
@jlurien It seems new state check is missing in the detailed steps below

jlurien replied:
Text rephrased

jlurien commented on documentation/API-Testing-Guidelines.md:
````diff
@@ -1,41 +1,288 @@
-Subproject_Repository/code/Test_definitions/file_name.feature
-for e.g. https://github.com/camaraproject/QualityOnDemand/blob/main/code/Test_definitions/QoD_API_Test.feature
-```
+Based on the decision taken in the Release Management and Commonalities working groups, and as documented in [API-Readiness-Checklist](https://github.com/camaraproject/Commonalities/blob/main/documentation/API-Readiness-Checklist.md), at least one Gherkin feature file will be added to the main subproject repo and this will fulfill the minimum criteria of readiness with respect to API test cases a ...
````
We can clarify that explicitly that, but the "API Readiness Checklist" applies to each API individually, so yes, each API which is part of a meta-release, regardless whether it is sharing a subproject with other APIs, must have its own test plan.

mdomale commented on documentation/API-Testing-Guidelines.md:
```diff
@@ -1,41 +1,288 @@
+
+For the explicit subscriptions model:
+
+* Validate that the subscribed events are received in the `notificationUrl`, with the right `notificationAuthToken`, for those situations specified in the API.
```
@jlurien  Consideration of new template  event-subscription-template.yaml needed in Testing-Guidelines as well

jlurien replied:
I don't get this comment, sorry. Could you please elaborate further? Which template and which considerations?

mdomale replied:
@jlurien https://github.com/camaraproject/Commonalities/pull/189 This was new template introduced & consideration of this for Testing guidelines will be useful cc @akoshunyadi

rartych replied:
@jlurien The template for explicit event subscription (like geofencing): https://github.com/camaraproject/Commonalities/blob/main/artifacts/camara-cloudevents/event-subscription-template.yaml

jlurien replied:
It should be aligned now after @PedroDiez's suggestions. Please double check

mdomale commented on documentation/API-Testing-Guidelines.md:
```diff
@@ -1,41 +1,288 @@
+### Error scenarios <a name="error-scenarios"></a>
+
+* All errors explicitly documented in the API spec must be covered by one or more dedicated error scenarios.
+  - In particular operations that expect a `device` object allowing one or more identifiers to be provided, must include scenarios to test all error cases defined. 
```
@jlurien Can you please clarify why only device object is considered explicitly? We think generic phrase would be fine for covering error scenarios.

jlurien replied:
Because recently errors for device processing have been discussed specifically, and new design guidelines also have a dedicated mention to device error management. But this do not exclude any other error scenario. Idea is to give an example of deep of tests that are expected,

jlurien replied:
added reference to the device error management section of the API design guidelines

mdomale commented on documentation/API-Testing-Guidelines.md:
```diff
@@ -1,41 +1,288 @@
+
+- APIs with a single feature file will use the api-name as file name: E.g. `location-verification.feature`
+- For APIs with several operations that split the test plan in one feature file per operation, recommendation is to use the `operationId` as file name: `createSession.feature`.
+- For APIs with several feature files, not split by operation, may use the api-name along some other description, e.g.: `qod-notifications.feature`
```
@jlurien  We think API name should be mandatory with some added description if needed .

jlurien replied:
Fine to me, I could suggest having always the api-name and, when the testing plan is composed of several files, that suffix, which can be operationId or other description.

jlurien replied:
Suggestion from @rartych applied

mdomale replied:
@jlurien  File still does not contain apiname-operationid/description in 2nd point  for Createsession.feature.
Kindly amend in the PR

rartych replied:
I think adding api-name in the second point brings consistency (filename always starts with api-name) , but can be helpful in the future with multiple API specifications in the same repository (distinguishing tests in case the same operationId is used).

> - For APIs with several operations that split the test plan in one feature file per operation, recommendation is to use the api-name along the `operationId` as file name: `device-identifier-retrieveType.feature`.

jlurien replied:
I agree. Applied

rartych commented on documentation/API-Testing-Guidelines.md:
```diff
@@ -1,41 +1,288 @@
+
+### Independent operations <a name="independent-operations"></a>
+
+Operations that execute an action or retrieve information from the operator, without modifying a resource, will be tested in its own dedicated feature file. 
```
```suggestion
Operations that execute an action or retrieve information from the operator, without modifying a resource, will be usually tested in its own dedicated feature file. 
```

rartych commented on documentation/API-Testing-Guidelines.md:
```diff
@@ -1,41 +1,288 @@
+For operations receiving an input request body with optional properties, several scenarios should be included, testing different sets of valid inputs.
+
+
+### CRUD operations <a name="crud-operations"></a>
```
```suggestion
### Resource modifying operations <a name="crud-operations"></a>
```

rartych commented on documentation/API-Testing-Guidelines.md:
```diff
@@ -1,41 +1,288 @@
+
+- APIs with a single feature file will use the api-name as file name: E.g. `location-verification.feature`
+- For APIs with several operations that split the test plan in one feature file per operation, recommendation is to use the `operationId` as file name: `createSession.feature`.
+- For APIs with several feature files, not split by operation, may use the api-name along some other description, e.g.: `qod-notifications.feature`
```
```suggestion
- For APIs with several feature files, not split by operation, should use the api-name along some other description, e.g.: `qod-notifications.feature`
```

jlurien replied:
I rephrased the sentence

PedroDiez reviewed (dismissed):
Some minor editorials to be reviewed
Fine at content level

LGTM in advance

jlurien commented:
I think I have incorporated all suggestions and feedback, please double check

PedroDiez reviewed (approved):
LGTM

mdomale commented:
@jlurien  Except above 1 comment PR looks fine to us .

jlurien commented:
Final comments addressed. Thanks for the feedback and corrections.

rartych reviewed (approved):
LGTM

patrice-conil reviewed (approved):
LGTM

mdomale commented:
@jlurien Thanks for accommodating comments .LGTM  now cc @akoshunyadi  @shilpa-padgaonkar

mdomale reviewed (approved):
LGTM thanks

PedroDiez reviewed (approved):
LGTM
